consensus/test/member_management_test.go

```
@@ -0,0 +1,526 @@
+	"github.com/clearmatics/autonity/crypto"
+	"github.com/clearmatics/autonity/ethclient"
+	"github.com/clearmatics/autonity/p2p/enode"
+	"github.com/stretchr/testify/require"
+	"math/big"
+	"strconv"
+	"sync"
+	"testing"
+)
+
+/*
+  In this file, it create 4 test cases which have similar work flow base on the local e2e test framework's main flow.
+
+  First it setup an autontiy network by according to the genesis hook function, then from the specific chain height, it
+  start to issue transactions via the transaction hook function specified for the target node, for example in the
+  addValidatorHook, it issues transaction to call autonity contract via operator account to add a new validator.
+
+  Then the test case verify the output from its finalAssert hook function on the specified height of the blockchain, for
+  example the addValidatorCheckerHook checks if the new validator is presented in the white list, and its stake balance
+  checked too, and finally it checks the total stake supply after the membership updates.
+
+  for the other cases in the file: add stake_holder, participants, or remove user, they follow the same work flow and
+  some rules to check the outputs.
+*/
+
+func TestMemberManagementAddNewValidator(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode")
+	}
+	once := sync.Once{}
+	// prepare chain operator
+	operatorKey, err := keygenerator.Next()
+	if err != nil {
+		t.Fatal(err)
+	}
+	operatorAddress := crypto.PubkeyToAddress(operatorKey.PublicKey)
+
+	newValidatorKey, err := keygenerator.Next()
+	require.NoError(t, err)
+
+	stakeBalance := new(big.Int).SetUint64(300)
+
+	addValidatorHook := func(validator *testNode, _ common.Address, _ common.Address) (bool, *types.Transaction, error) { //nolint
+		if validator.lastBlock <= 3 {
+			return true, nil, nil
+		}
+		once.Do(func() {
+			contract, err := autonityInstance(validator.rpcPort)
+			if err != nil {
+				t.Fatal(err)
```

@Jason-Zhangxin-Chen, you can't use `t.Fatal` inside any beforeHooks, afterHooks or sendTransactionHooks, because they are all run in a different goroutine to the main test goroutine.

You need to return the errors in these cases.

---

There is already a error return value, so please use that to return the error to the caller

---

Thank you for the findings, @piersy @aneequesafdar , I fix them in my lastest commit today.